ts/session/medium_group/index.ts

```diff
@@ -39,56 +59,844 @@ async function createSenderKeysForMembers(
+    return;
+  }
+
+  const convo: ConversationModel = maybeConvo;
+
+  const now = Date.now();
+
+  convo.set({ left: true });
+  await convo.commit();
+
+  const dbMessage = await convo.addMessage({
+    group_update: { left: 'You' },
+    conversationId: groupId,
+    type: 'outgoing',
+    sent_at: now,
+    received_at: now,
+  });
+
+  const updateParams = {
+    timestamp: Date.now(),
+    identifier: dbMessage.id,
+    groupId,
+  };
+
+  const message = new MediumGroupQuitMessage(updateParams);
+
+  await sendToMembers(groupId, message, dbMessage);
+}
+
+// Just a container to store two named list of keys
+interface SenderKeysContainer {
+  newKeys: Array<RatchetState>;
+  existingKeys: Array<RatchetState>;
+}
+
+// Load all known keys for all members (or only for select devices if specified)
+async function getExistingSenderKeysForGroup(
+  groupId: string,
+  devices: Array<PubKey>
+): Promise<Array<RatchetState>> {
+  const maybeKeys = await Promise.all(
+    devices.map(async device => {
+      const maybeKey = await getChainKey(groupId, device.key);
+
+      if (!maybeKey) {
+        return null;
+      } else {
+        const { chainKey, keyIdx } = maybeKey;
+        const pubKeyBin = new Uint8Array(fromHex(device.key));
+        return {
+          chainKey: new Uint8Array(chainKey),
+          keyIdx,
+          pubKey: pubKeyBin,
+        };
+      }
+    })
+  );
+
+  return maybeKeys.filter(d => d !== null).map(d => d as RatchetState);
+}
+
+// Create all sender keys based on the changes in
+// the group's composition
+async function getOrCreateSenderKeysForUpdate(
+  groupId: string,
+  members: Array<string>,
+  changes: MemberChanges
+): Promise<SenderKeysContainer> {
+  // 1. Create sender keys for every joining member
+  const joining = changes.joiningMembers || [];
+  const leaving = changes.leavingMembers || [];
+
+  let newKeys = await createSenderKeysForMembers(groupId, joining);
+
+  // 2. Get ratchet states for existing members
+
+  const existingMembers = _.difference(members, joining);
+  // get all devices for members
+
+  const allDevices = _.flatten(
+    await Promise.all(
+      existingMembers.map(m => MultiDeviceProtocol.getAllDevices(m))
+    )
+  );
+
+  let existingKeys: Array<RatchetState> = [];
+
+  if (leaving.length > 0) {
+    // If we have leaving members, we have to re-generate ratchet
+    // keys for existing members
+    const otherKeys = await Promise.all(
+      allDevices.map(async device => {
+        return createSenderKeyForGroup(groupId, PubKey.cast(device));
+      })
+    );
+
+    newKeys = _.union(newKeys, otherKeys);
+  } else {
+    // We can reuse existing keys
+    existingKeys = await getExistingSenderKeysForGroup(groupId, allDevices);
+  }
+
+  return { existingKeys, newKeys };
+}
+
+async function getGroupSecretKey(groupId: string): Promise<Uint8Array> {
+  const groupIdentity = await window.Signal.Data.getIdentityKeyById(groupId);
```

Comment:
Shouldn't we avoid using window when we can import the file as ts import? Not sure we can for this one

Reply:
Updated.